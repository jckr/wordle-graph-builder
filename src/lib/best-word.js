import { scoreWord, evaluateScore } from './score';
/**
 * @param {string[]} solutions
 * @param {string[]} validWords
 */
export function findBestWord(solutions, validWords) {
  const results = {};
  let bestScoreValue = Infinity;
  let bestScore = {};
  let bestWord = '';
  for (const word of validWords) {
    const wordScore = scoreWord(word, solutions);
    const scoreValue = evaluateScore(wordScore);
    results[word] = { wordScore, scoreValue };
    if (scoreValue < bestScoreValue) {
      bestWord = word;
      bestScoreValue = scoreValue;
      bestScore = wordScore;
    }
  }
  return { bestWord, bestScoreValue, bestScore, results };
}