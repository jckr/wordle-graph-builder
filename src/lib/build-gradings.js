import { readFileSync } from 'node:fs';

/**
 *  @typedef {{[key: string]: number[]}} Grades
 *  @typedef {{[key: string]: Set<number>}} GradeSet
 * 
 */

/**
 * @param {string} grade
 * @returns {number} */
 export function gradeToNumber(grade) {
  const letters = grade.split('');
  return letters.reverse().reduce((prev, letter, index) => {
    if (letter === 'P') {
      return prev + 1000 + 10 * index;
    }
    if (letter === 'p') {
      return prev + 100 + index;
    }
    return prev;
  }, 0);
};
/**
 * For a given solution and a given word, finds the score that entering this word would give.
 * This is used to prepare the gradings file. 
 * @param {string} solution
 * @param {string} word
 * @returns {number} */
export function gradeSingleWord(solution, word) {
  const solutionLetters = solution.split('');
  const wordLetters = word.split('');
  let grade = '';
  for (let i = 0; i < solutionLetters.length; i++) {
    const solutionLetter = solutionLetters[i];
    const wordLetter = wordLetters[i];
    if (solutionLetter === wordLetter) {
      grade += 'P';
    } else if (solutionLetters.includes(wordLetter)) {
      grade += 'p';
    } else {
      grade += 'A';
    }
  }
  return gradeToNumber(grade);
}

/**
 * Prepares the gradings file. This contains the scores for all words for all solutions.
 * @param {string[]} solutions
 * @param {string[]} words
 * @returns {Grades[]} */
 function gradeAllWords(solutions, words) {
  const allGrades = [];
  for (let solutionsIndex = 0; solutionsIndex < solutions.length; solutionsIndex++) {
    const solution = solutions[solutionsIndex];
    /** @type {Grades} */const grades = {};
    for (let wordIndex = 0; wordIndex < words.length; wordIndex++) {
      const word = words[wordIndex];
      const grade = gradeSingleWord(solution, word);
      grades[grade] = grades[grade] || [];
      grades[grade].push(wordIndex);
    }
    allGrades.push(grades);
  }
  return allGrades;
};

/**
 * Modifies the gradings data from array to set for faster lookup.
 * @param {Grades[]} gradings
 * @returns {GradeSet[]}
 */
 export function settify(gradings) {
  return gradings.map(grades => {
    /** @type {GradeSet} */const gradeSet = {};
    for (const grade in grades) {
      gradeSet[grade] = new Set(grades[grade]);
    }
    return gradeSet;
  });
}

/**
 * Retrieves the gradings, either from a file or by calculating them from words and solutions.
 * @param {string[]} solutions 
 * @param {string[]} words 
 * @param {string?} gradingsFile 
 * @returns {Grades[]}
 */
 export function getGradings(solutions, words, gradingsFile) {
  if (gradingsFile) {
    const gradingsRaw = readFileSync(gradingsFile, 'utf8');
    return JSON.parse(gradingsRaw);
  }  
  return gradeAllWords(solutions, words);;
}