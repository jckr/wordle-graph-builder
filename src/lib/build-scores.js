import { readFileSync } from 'node:fs';
/**
 * @typedef {number[][]} score
 */

/**
 * @param {string} grade
 * @returns {number} */
function gradeToNumber(grade) {
  const letters = grade.split('');
  return letters.reverse().reduce((prev, letter, index) => {
    if (letter === 'P') {
      return prev + 1000 + 10 * index;
    }
    if (letter === 'p') {
      return prev + 100 + index;
    }
    return prev;
  }, 0);
};
/**
 * For a given solution and a given word, finds the score that entering this word would give.
 * This is used to prepare the gradings file. 
 * @param {string} solution
 * @param {string} word
 * @returns {number} */
function gradeSingleWord(solution, word) {
  const solutionLetters = solution.split('');
  const wordLetters = word.split('');
  let grade = '';
  for (let i = 0; i < solutionLetters.length; i++) {
    const solutionLetter = solutionLetters[i];
    const wordLetter = wordLetters[i];
    if (solutionLetter === wordLetter) {
      grade += 'P';
    } else if (solutionLetters.includes(wordLetter)) {
      grade += 'p';
    } else {
      grade += 'A';
    }
  }
  return gradeToNumber(grade);
}

/**
 * Prepares the gradings file. This contains the scores for all words for all solutions.
 * @param {string[]} solutions
 * @param {string[]} words
 * @returns {score} */
function scoreAllWords(solutions, words) {
  const allScores = [];
  console.log('scoring');
  for (let solutionsIndex = 0; solutionsIndex < solutions.length; solutionsIndex++) {
    process.stdout.write('.');
    const solution = solutions[solutionsIndex];
    const scores = [];
    for (let wordIndex = 0; wordIndex < words.length; wordIndex++) {
      const word = words[wordIndex];
      const grade = gradeSingleWord(solution, word);
      scores.push(grade);
    }
    allScores.push(scores);
  }
  console.log('done');
  return allScores;
};

/**
 * Retrieves the gradings, either from a file or by calculating them from words and solutions.
 * @param {string[]} solutions 
 * @param {string[]} words 
 * @param {string?} scoringFile 
 * @returns {score}
 */
 export function getScores(solutions, words, scoringFile) {
  if (scoringFile) {
    const gradingsRaw = readFileSync(scoringFile, 'utf8');
    return JSON.parse(gradingsRaw);
  }  
  return scoreAllWords(solutions, words);;
}