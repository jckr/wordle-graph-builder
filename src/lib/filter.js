export class Filter {
  correct = {};
  present = {};
  absent = new Set();
  /**
   * 
   * @param {string} letter 
   * @param {number} position 
   */
  #addCorrect(letter, position) {
    if (this.absent.has(letter)) {
      throw(`Letter ${letter} can't be correct and absent`);
    }
    if (this.present[letter]?.has(position)) {
      throw(`Letter ${letter} can't be correct and present at same position`);
    }
    if (this.correct[letter]?.has(position)) {
      return;
    }
    if (Object.keys(this.correct).length >= 5) {
      throw(`Can't have more than 5 correct letters`);
    }
    this.correct[letter] = this.correct[letter] || new Set();
    this.correct[letter].add(position);
  }
  /**
   * 
   * @param {string} letter 
   */
  #addAbsent(letter) {
    if (this.correct[letter]) {
      throw(`Letter ${letter} can't be correct and absent`);
    }
    if (this.present[letter]) {
      throw(`Letter ${letter} can't be present and absent`);
    }
    this.absent.add(letter);
  }
  /**
   * 
   * @param {string} letter 
   * @param {number} position 
   */
  #addPresent(letter, position) {
    if (this.absent.has(letter)) {
      throw(`Letter ${letter} can't be present and absent`);
    }
    if (this.correct[letter]?.has(position)) {
      throw(`Letter ${letter} can't be correct and present at same position`);
    }
    if (this.present[letter]?.has(position)) {
      return;
    }
    if (Object.keys(this.present).length >= 5) {
      throw(`Can't have more than 5 present letters`);
    }
    this.present[letter] = this.present[letter] || new Set();
    this.present[letter].add(position);
    if (Object.keys(this.present).length === 5) {
      'abcdefghijklmnopqrstuvwxyz'.split('').forEach(letter => {
        if (!this.present[letter] && !this.correct[letter]) {
          this.#addAbsent(letter);
        }
      });
    }
  }
  /**
   * 
   * @param {string} word 
   * @param {string} grade
   */
  add(word, grade) {
    if (word.length !== 5) {
      throw('Word must be 5 letters');
    }
    if (grade.length !== 5) {
      throw('Grade must be 5 letters');
    }
    const letters = word.split('');
    for (let p = 0; p < grade.length; p++) {
      const g = grade[p];
      const letter = letters[p];
      if (g === 'C') {
        this.#addCorrect(letter, p);
      } else if (g === 'P') {
        this.#addPresent(letter, p);
      } else if (g === 'A') {
        this.#addAbsent(letter);
      } else {
        throw(`Unknown grade ${g}`);
      }
    }
  }
  derive(word, grade) {
    const f = new Filter();
    for (const a of this.absent) {
      f.#addAbsent(a);
    }
    for (const c in this.correct) {
      for (const i of this.correct[c]) {
        f.#addCorrect(c, i);
      }
    }
    for (const p in this.present) {
      for (const i of this.present[p]) {
        f.#addPresent(p, i);
      }
    }
    f.add(word, grade);
    return f;
  }
}