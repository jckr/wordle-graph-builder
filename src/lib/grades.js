export default [
  'AAAAA',
  'AAAAP',
  'AAAPA',
  'AAPAA',
  'APAAA',
  'PAAAA',
  'AAAPP',
  'AAPAP',
  'AAPPA',
  'APAAP',
  'APAPA',
  'APPAA',
  'PAAAP',
  'PAAPA',
  'PAPAA',
  'PPAAA',
  'AAPPP',
  'APAPP',
  'APPAP',
  'APPPA',
  'PAAPP',
  'PAPAP',
  'PAPPA',
  'PPAAP',
  'PPAPA',
  'PPPAA',
  'APPPP',
  'PAPPP',
  'PPAPP',
  'PPPAP',
  'PPPPA',
  'PPPPP',
  'AAAAC',
  'AAACA',
  'AACAA',
  'ACAAA',
  'CAAAA',
  'AAAPC',
  'AAPAC',
  'APAAC',
  'PAAAC',
  'AAACP',
  'AAPCA',
  'APACA',
  'PAACA',
  'AACAP',
  'AACPA',
  'APCAA',
  'PACAA',
  'ACAAP',
  'ACAPA',
  'ACPAA',
  'PCAAA',
  'CAAAP',
  'CAAPA',
  'CAPAA',
  'CPAAA',
  'AAPPC',
  'APAPC',
  'APPAC',
  'PAAPC',
  'PAPAC',
  'PPAAC',
  'AAPCP',
  'APACP',
  'APPCA',
  'PAACP',
  'PAPCA',
  'PPACA',
  'AACPP',
  'APCAP',
  'APCPA',
  'PACAP',
  'PACPA',
  'PPCAA',
  'ACAPP',
  'ACPAP',
  'ACPPA',
  'PCAAP',
  'PCAPA',
  'PCPAA',
  'CAAPP',
  'CAPAP',
  'CAPPA',
  'CPAAP',
  'CPAPA',
  'CPPAA',
  'APPPC',
  'PAPPC',
  'PPAPC',
  'PPPAC',
  'APPCP',
  'PAPCP',
  'PPACP',
  'PPPCA',
  'APCPP',
  'PACPP',
  'PPCAP',
  'PPCPA',
  'ACPPP',
  'PCAPP',
  'PCPAP',
  'PCPPA',
  'CAPPP',
  'CPAPP',
  'CPPAP',
  'CPPPA',
  'PPPPC',
  'PPPCP',
  'PPCPP',
  'PCPPP',
  'CPPPP',
  'AAACC',
  'AACAC',
  'AACCA',
  'ACAAC',
  'ACACA',
  'ACCAA',
  'CAAAC',
  'CAACA',
  'CACAA',
  'CCAAA',
  'AAPCC',
  'APACC',
  'PAACC',
  'AACPC',
  'APCAC',
  'PACAC',
  'AACCP',
  'APCCA',
  'PACCA',
  'ACAPC',
  'ACPAC',
  'PCAAC',
  'ACACP',
  'ACPCA',
  'PCACA',
  'ACCAP',
  'ACCPA',
  'PCCAA',
  'CAAPC',
  'CAPAC',
  'CPAAC',
  'CAACP',
  'CAPCA',
  'CPACA',
  'CACAP',
  'CACPA',
  'CPCAA',
  'CCAAP',
  'CCAPA',
  'CCPAA',
  'APPCC',
  'PAPCC',
  'PPACC',
  'APCPC',
  'PACPC',
  'PPCAC',
  'APCCP',
  'PACCP',
  'PPCCA',
  'ACPPC',
  'PCAPC',
  'PCPAC',
  'ACPCP',
  'PCACP',
  'PCPCA',
  'ACCPP',
  'PCCAP',
  'PCCPA',
  'CAPPC',
  'CPAPC',
  'CPPAC',
  'CAPCP',
  'CPACP',
  'CPPCA',
  'CACPP',
  'CPCAP',
  'CPCPA',
  'CCAPP',
  'CCPAP',
  'CCPPA',
  'PPPCC',
  'PPCPC',
  'PPCCP',
  'PCPPC',
  'PCPCP',
  'PCCPP',
  'CPPPC',
  'CPPCP',
  'CPCPP',
  'CCPPP',
  'AACCC',
  'ACACC',
  'ACCAC',
  'ACCCA',
  'CAACC',
  'CACAC',
  'CACCA',
  'CCAAC',
  'CCACA',
  'CCCAA',
  'APCCC',
  'PACCC',
  'ACPCC',
  'PCACC',
  'ACCPC',
  'PCCAC',
  'ACCCP',
  'PCCCA',
  'CAPCC',
  'CPACC',
  'CACPC',
  'CPCAC',
  'CACCP',
  'CPCCA',
  'CCAPC',
  'CCPAC',
  'CCACP',
  'CCPCA',
  'CCCAP',
  'CCCPA',
  'PPCCC',
  'PCPCC',
  'PCCPC',
  'PCCCP',
  'CPPCC',
  'CPCPC',
  'CPCCP',
  'CCPPC',
  'CCPCP',
  'CCCPP',
  'ACCCC',
  'CACCC',
  'CCACC',
  'CCCAC',
  'CCCCA',
  'PCCCC',
  'CPCCC',
  'CCPCC',
  'CCCPC',
  'CCCCP',
  'CCCCC'
];