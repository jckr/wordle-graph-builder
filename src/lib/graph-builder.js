import { writeFileSync } from 'node:fs';
import {loadData} from './load-data.js';
import {gradeSingleWord,  gradeToNumber } from './build-gradings.js';
import {getScores} from './build-scores.js';

/**
 *  @typedef {{[key: string]: number[]}} Grades
 *  @typedef {{[key: string]: Set<number>}} GradeSet
 *  @typedef {{grade: number, solution: number}} Solution
 *  @typedef {{grade: number, move: number, children: (Solution|Branch)[]}} Branch
 *  @typedef {{root: number, children: (Solution|Branch)[]}} Graph
 */

/**
 * Given a word (as a number), and a set of solutions (also as numbers), groups these solutions by grade
 * @param {number} wordIndex 
 * @param {number[]} possibleSolutions 
 * @param {GradeSet[]} gradings
 * @returns {Grades} 
 */

function groupSolutions(wordIndex, possibleSolutions, gradings) {
  /** @type {Grades} */const groups = {};
  for (const solutionIndex of possibleSolutions) {
    
    const grade = gradings[solutionIndex];
    const possibleGrades = Object.entries(grade);
    for (const [grade, words] of possibleGrades) {
      if (words.has(wordIndex)) {
        groups[grade] = groups[grade] || [];
        groups[grade].push(solutionIndex);
      }
    }
  }
  return groups;
}
/**
 * This is a super basic, almost trash heuristic function - we just take the
 * first possible solution. Normally, we would go through all the possible words
 * (hence the nbWords parameter) and compare the characteristics obtained from 
 * getting a group for each of them given the possible solutions. 
 * @param {GraphBuilder} graphBuilder
 * @param {number[]} possibleSolutions
 * @returns {{bestMove: number, groups: Grades}}
*/
function basicHeuristic(graphBuilder, possibleSolutions) {
  const bestMove = possibleSolutions[0];
  const groups = graphBuilder.groupSolutions(bestMove, possibleSolutions);
  return {bestMove, groups};
}


/**
 * This heuristic favors an aggressive approach. For each group of possible solutions,
 * we choose the word which has the highest chance of finding the solution in just one
 * extra move. 
 * @param {GraphBuilder} graphBuilder
 * @param {number[]} possibleSolutions
 * @returns {{bestMove: number, groups: Grades}}
*/
function fastHeuristic(graphBuilder, possibleSolutions) {
  let topScore = 0;
  let bestMove = 0;  
  /** @type Grades */ let bestGroups = {};
  for (let wordIndex = 0; wordIndex < graphBuilder.nbWords; wordIndex++) {
    const groups = graphBuilder.groupSolutions(wordIndex, possibleSolutions);
    const score = Object.values(groups).filter(group => group.length < 3).length;
    if (score > topScore) {
      topScore = score;
      bestMove = wordIndex;
      bestGroups = groups;
    }
    wordIndex++;
  }
  return {bestMove, groups: bestGroups};
}

class GraphBuilder {
  moveCounter = 0;
  /**
   * @param {number[][]} scores
   * @param {function} heuristic
   * @param {string[]} solutions
   * @param {string[]} words
   * @param {number} wordIndex
   */
  constructor(scores, heuristic, solutions, words, wordIndex) {
    this.scores = scores;
    this.heuristic = heuristic || basicHeuristic;
    this.nbSolutions = solutions.length;
    this.nbWords = words.length;
    this.solutions = solutions;
    this.words = words;
    this.wordIndex = wordIndex;
  }
  /**
   * @param {number} wordIndex
   * @param {number[]} possibleSolutions 
   * @returns {Grades}
   */
  groupSolutions(wordIndex, possibleSolutions) {
    /** @type {Grades} */const groups = {};
    for (const solutionIndex of possibleSolutions) {
      const grade = this.scores[solutionIndex][wordIndex];
      groups[grade] = groups[grade] || [];
      groups[grade].push(solutionIndex);
    }
    return groups;
  }
  /**
   * @param {number} grade
   * @param {number[]} possibleSolutions
   * @returns {Branch|Solution}
  */
  nextMove(grade, possibleSolutions) {
    this.moveCounter++;
    if (this.moveCounter % 100 === 0) {
      process.stdout.write('.');
    }
    // if there is just one possible solution, we know this is the best move. 
    if (possibleSolutions.length === 1) {
      const solution = possibleSolutions[0];
      return {
        grade,
        solution
      };
    }
    if (possibleSolutions.length === 2) {
      const [firstSolution, secondSolution] = possibleSolutions;
      return {
        grade,
        move: firstSolution,
        children: [{
          grade: this.scores[firstSolution][firstSolution],
          solution: firstSolution,
        },
        {
          grade: this.scores[secondSolution][firstSolution],
          solution: secondSolution
        }]
      };
    }
    // if there are more than two possible solutions, we use the heuristic function
    // to determine the best move. 

    const {bestMove, groups} = this.heuristic(this, possibleSolutions);
    const children = Object.entries(groups)
      // highest grade to lowest grade
      .sort((a, b) => Number(b[0]) - Number(a[0]))
      .map(([grade, solutions]) => this.nextMove(
        Number(grade),
        solutions, 
      ));
      
    return {
      move: bestMove,
      grade,
      children
    };  
  }
  /**
   * 
   * @returns {string}
   */
  build() {
    const allSolutions = [...Array(this.nbSolutions).keys()];
    const groupsForRoot = this.groupSolutions(this.wordIndex, allSolutions);
    const start = new Date();
    console.log(start, 'Building graph...');
    /** @type {Graph} */ const graph = {
      root: this.wordIndex,
      children: Object.entries(groupsForRoot).map(([grade, remainingSolutions]) => {
        return this.nextMove(Number(grade), remainingSolutions, 1);
      })
    };
    // removing the quotes from the output to save space
    const end = new Date();
    console.log(end, `done. ${end.valueOf() - start.valueOf()}ms`);
    return `export default ${JSON.stringify(graph).replaceAll('"', '')};`;
  }
}

/**
 * @param {number} startWord
 * @param {number} endWord
 * @param {string} prefix
 * @param {string?} scoringFile 
 * @param {boolean?} save
 */
function buildGraph(startWord = 0, endWord = Infinity, prefix, scoringFile, save) {
  const {solutions, allWords} = loadData('../data/words.json');

  const scores = getScores(solutions, allWords, scoringFile);
  if (!scoringFile) {
    writeFileSync('../data/scoring.json', JSON.stringify(scores));
  }
  for (let wordIndex = startWord; wordIndex < endWord && wordIndex < allWords.length; wordIndex++) {
    console.log('now working on', wordIndex);
    const graphBuilder = new GraphBuilder(scores, fastHeuristic, solutions, allWords, wordIndex);
    const graph = graphBuilder.build();
    if (save) {
      writeFileSync(`../data/graphs/${prefix}/${allWords[wordIndex]}.js`, graph);
    } else {
      console.log(graph);
    }
  }
}

buildGraph(Number(process.argv[2]), Number(process.argv[3]), 'fast', '../data/scores.json', true);