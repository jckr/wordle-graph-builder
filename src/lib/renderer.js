import { numberToGrade } from "./build-scores";
/**
*  @typedef {number[]} Solution
*  @typedef {(number|Solution|Branch)[]} Branch
*  @typedef {(number|Solution|Branch)[]} Graph
*/
/**
 * @param {Graph} graphAsArray
 * @param {string[]} wordList
 */
export function renderer(graphAsArray, wordList) {
  const tree = renderBranch(graphAsArray, wordList, 1);
  return computeAngle(tree, 0, 2 * Math.PI);
}

/**
 * @param {Solution} solution 
 * @param {string[]} wordList 
 */
function renderSolution(solution, wordList, nbMoves) {
  const canBeSolution = solution[0] === 242;
  const movesTally = [...Array(nbMoves + (canBeSolution ? 1 : 2)).keys()].map(d => 0);
  movesTally[nbMoves + (canBeSolution ? 0 : 1)] = 1;
  return {
    canBeSolution,
    movesTally,
    nbMoves,
    grade: numberToGrade(solution[0]),
    move: wordList[solution[1]],
    size: 1,
    nbSolutions: 1
  };
}
/**
 * @param {Branch} branch
 * @param {string[]} wordList 
 */
function renderBranch(branch, wordList, nbMoves) {
  if (branch.length === 2) {
    return renderSolution(branch, wordList, nbMoves);
  };
  const childrenWithoutSolution = branch[2].filter(child => child[0] !== 242);
  const canBeSolution = childrenWithoutSolution.length < branch[2].length;

  const children = childrenWithoutSolution.map(child => renderBranch(child, wordList, nbMoves + 1));
  const size = children.reduce((acc, child) => acc + child.size, 0);
  const nbSolutions = size + (canBeSolution ? 1 : 0);
  const movesTally = children.reduce((acc, child) => {
    child.movesTally.forEach((childMoveTally, index) => {
      acc[index] = (acc[index] || 0) + childMoveTally;
    });
    return acc;
  }, []);
  if (canBeSolution) {
    movesTally[nbMoves] = (movesTally[nbMoves] || 0) + 1;
  }
  return {
    canBeSolution,
    children,
    nbMoves,
    grade: numberToGrade(branch[0]),
    movesTally,
    move: wordList[branch[1]],
    nbSolutions,
    size
  };
}

function computeAngle(tree, minAngle, maxAngle) {
  let runningAngle = minAngle;
  return {
    ...tree,
    minAngle,
    maxAngle,
    ...(tree.children ? {children: tree.children.map((child, index) => {
      const relativeSize = (maxAngle - minAngle) * child.size / tree.size;
      return computeAngle(child, runningAngle, runningAngle += relativeSize);
    })} : {})
  };
}