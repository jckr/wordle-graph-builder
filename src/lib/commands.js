import { existsSync, writeFileSync } from 'node:fs';
import {GraphBuilder} from './graph-builder.js';
import {loadData} from './load-data.js';
import {getScores} from './build-scores.js';
import {getHeuristic} from './heuristics.js';

/**
 * @param {number[]} wordList
 * @param {string[]} words
 * @param {string[]} solutions
 * @param {number[][]} scores
 * @param {string} prefix
 * @param {function} heuristic
 * @param {boolean} save
 */
 function buildGraph(wordList, words, solutions, scores, prefix, heuristic, save) {
  for (const wordIndex of wordList) {
    console.log('now working on', words[wordIndex]);
    const graphBuilder = new GraphBuilder(scores, heuristic, solutions, words, wordIndex);
    const graph = graphBuilder.build();
    if (save) {
      console.log('saving.');
      writeFileSync(`./src/data/graphs/${prefix}/${words[wordIndex]}.json`, graph);
    } else {
      console.log(graph);
    }
  }
}

/**
 * @param {number} start
 * @param {number} end
 * @param {string?} mode
 * @param {boolean?} save
*/ 
export function buildRange(start, end, mode = 'fast', save) {
  const {solutions, allWords} = loadData('./src/data/words.json');
  const scores = getScores(solutions, allWords, './src/data/scores.json');
  if (!existsSync('./src/data/scores.json')) {
    writeFileSync('./src/data/scores.json', JSON.stringify(scores));
  }
  const wordList = [...Array(end).keys()].slice(start);
  const {prefix, heuristic} = getHeuristic(mode);
  buildGraph(wordList, allWords, solutions, scores, prefix, heuristic, save ?? true);
}

/**
 * @param {string?} mode 
 * @param {boolean?} save
*/ 
export function buildPopular(mode = 'fast', save) {
  const {solutions, allWords, popularWords} = loadData('./src/data/words.json');
  const wordList = popularWords.map(word => allWords.indexOf(word));
  const scores = getScores(solutions, allWords, './src/data/scores.json');
  if (!existsSync('./src/data/scores.json')) {
    writeFileSync('./src/data/scores.json', JSON.stringify(scores));
  }
  const {prefix, heuristic} = getHeuristic(mode);
  buildGraph(wordList, allWords, solutions, scores, prefix, heuristic, save ?? true);
}
