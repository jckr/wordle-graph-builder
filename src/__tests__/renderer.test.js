import { renderer } from '../lib/renderer';
import { readFileSync } from 'node:fs';

describe('renderer works', () => {
  const {solutions, validWords} = JSON.parse(readFileSync('src/data/words.json', 'utf8'));
  const wordList = solutions.concat(validWords);
  const graphAsArray = JSON.parse(readFileSync('src/__tests__/testdata/cigar.json', 'utf8'));
  test('rendering cigar', () => {
    const tree = renderer(graphAsArray, wordList);
    expect(tree.movesTally).toMatchSnapshot();
  });
})