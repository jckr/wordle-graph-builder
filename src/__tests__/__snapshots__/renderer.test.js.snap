// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderer works rendering cigar 1`] = `
[
  0,
  1,
  35,
  319,
  627,
  1191,
  136,
]
`;
